prov-api/helpers/prov_connector.py

```diff
@@ -0,0 +1,1597 @@
+
+Key components: - Neo4j connection and query execution - Graph structure
+representations (Node, NodeLink, NodeGraph) - Graph querying and traversal -
+Graph comparison and diffing - Graph building and modification
+
+Note: This module assumes the use of a Neo4j database and requires appropriate
+configuration and credentials to connect to the database.
+
+TODO: Consider alternative ways of embedding record IDs which will scale more as
+record count increases.
+"""
+import prov.model as prov  # type: ignore
+from typing import Dict, Callable, List
+from enum import Enum, auto
+from pydantic import BaseModel, Field, ValidationError
+from typing import Dict, List, Set, Any, Tuple, Callable
+from enum import Enum
+from config import Config
+from ProvenaInterfaces.RegistryModels import ItemSubType
+from ProvenaInterfaces.ProvenanceModels import *
+from ProvenaInterfaces.RegistryModels import *
+from typing import List, Any, Dict
+from fastapi import HTTPException
+from neo4j import GraphDatabase, Record, graph, basic_auth  # type: ignore
+from config import Config
+from typing import Optional, List, Any, Dict, Tuple, Callable
+import networkx  # type: ignore
+from dependencies.dependencies import secret_cache
+from helpers.keycloak_helpers import retrieve_secret_value
+
+# This is the property on which the principal id for the record is stored
+IDENTIFIER_TAG = "id"
+
+
+def produce_attribute_set(
+    item_category: ItemCategory,
+    item_subtype: ItemSubType,
+    id: str
+) -> Dict[str, Any]:
+    """    produce_attribute_set
+        Produces the set of attributes to add to the python-prov object.
+
+        This does NOT inform the attributes stored in the prov graph. Rather,
+        just the prov-o serialisation.
+
+        Arguments
+        ----------
+        item_category : ItemCategory
+            The category of the item
+        item_subtype : ItemSubType
+            The sub type of the item
+        id : str
+            The handle ID of the item
+
+        Returns
+        -------
+         : Dict[str, Any]
+            A set of attributes to add to the item
+
+        See Also (optional)
+        --------
+
+        Examples (optional)
+        --------
+    """
+    return {
+        'id': id,
+        'item_category': item_category.value,
+        'item_subtype': item_subtype.value
+    }
+
+
+def get_credentials(config: Config) -> Tuple[str, str]:
+    """
+    Retrieve Neo4j database credentials based on the provided configuration.
+
+    This function checks for test credentials first, and if not available,
+    retrieves credentials from a secret store.
+
+    Args:
+        config (Config): Configuration object containing credential information.
+
+    Returns:
+        Tuple[str, str]: A tuple containing the username and password for Neo4j.
+
+    Raises:
+        AssertionError: If the secret value is not in the expected format.
+    """
+    if config.neo4j_test_password and config.neo4j_test_username:
```

Comment:
cool - i just looked in the config object. I think it would be good if there was a validator to ensure that if a test password is provided a test username must also be too, and vice versa. This would catch if a dev only supplies a test username thinking test funcitonality would be performed but that wouldn't happen with this AND statement.

Reply:
This code already existed and was just moved, but could be good to enhance in the future